src/inc/AppxManifestObject.hpp

```diff
@@ -215,8 +215,34 @@ namespace MSIX {
         std::string m_name;
     };
 
+    class AppxManifestOptionalPackageInfo final : public ComClass<AppxManifestOptionalPackageInfo, IAppxManifestOptionalPackageInfo>
+    {
+    public:
+        AppxManifestOptionalPackageInfo(std::string mainPackageName) :
+            m_mainPackageName(mainPackageName)
+        {}
+
+        // IAppxManifestOptionalPackageInfo
+        HRESULT STDMETHODCALLTYPE GetIsOptionalPackage(BOOL* isOptionalPackage) noexcept override try
+        {
+            *isOptionalPackage = !m_mainPackageName.empty();
+            return static_cast<HRESULT>(Error::OK);
+        } CATCH_RETURN();
+
+        HRESULT STDMETHODCALLTYPE GetMainPackageName(LPWSTR* mainPackageName) noexcept override try
+        {
+            ThrowErrorIf(Error::InvalidParameter, (mainPackageName == nullptr || *mainPackageName != nullptr), "bad pointer");
+            std::wstring utf16MainPackageName = utf8_to_wstring(m_mainPackageName);
+            *mainPackageName = &utf16MainPackageName[0];

```

Comment:
>*mainPackageName = &utf16MainPackageName[0]; [](start = 12, length = 44)

This is not the way to marshall out a string. See MarshalOutStringUtf8/MarshalOutString